core/includes/form.inc

```diff
@@ -4889,6 +4906,97 @@ function _form_set_class(&$element, $class = array()) {
   }
 }
 
+/**
+ * Checks the number of input form elements to ensure that the PHP
+ * max_input_vars limit is not exceeded.
+ *
+ * @param array $count
+ *   The number of elements in the form.
+ * @param string $error_level
+ *   The error level for the message displayed on screen if the input vars value
+ *   has been exceeded.
+ *
+ * @see form_pre_render()
+ */
+function _form_check_input_vars($count, $error_level = 'warning') {
+  // Determine the limit on input vars for this server configuration.
+  $limit = ini_get('max_input_vars');
+  if ($limit) {
+    // Leave a 5% buffer for ajax elements or similar.
+    if ($count > $limit * 0.95) {
+      $substitutions = array(
+        '@count' => $count,
+        '@limit' => $limit,
+      );
+
+      // Translatable strings.
+      $warning = t('This form may be too long to work properly.');
+      $notice = t('The last form processed was longer than this system can safely support.');
+      $please_increase = t('This form contains @count input elements. PHP max_input_vars allows only @limit and should be increased.', $substitutions);
+      $data_loss = t('The form has been submitted, but data may heve been lost.');
+
+      switch ($error_level) {
+        case 'warning':
+          if (user_access('administer site configuration')) {
+            $message = $please_increase;
+          }
+          else {
+            $message = $warning;
+          }
+          break;
+
+        case 'error':
+          if (user_access('administer site configuration')) {
+            $message = $data_loss . ' ' . $please_increase;
+          }
+          else {
+            $message = $notice . ' ' . $data_loss;
+          }
+          break;
+      }
+
+      // Remove existing messages that could be confusing.
+      $set_new_message = TRUE;
+      if (!empty($_SESSION['messages'])) {
+
+        $current_path = current_path();
+        foreach ($_SESSION['messages'] as $level => $set) {
+          foreach ($_SESSION['messages'][$level] as $key => $previous_message) {
+            // Look for a specific message on the permissions page.
+            if ($current_path == 'admin/config/people/permissions' && $previous_message == t('The changes have been saved.')) {
+              unset($_SESSION['messages'][$level][$key]);
+            }
+            elseif (strstr($current_path, 'admin/structure/taxonomy') && $previous_message == t('The configuration options have been saved.')) {
+              unset($_SESSION['messages'][$level][$key]);
+            }
+            elseif (strstr($current_path, 'admin/structure/menu') && $previous_message == t('Your configuration has been saved.')) {
+              unset($_SESSION['messages'][$level][$key]);
+            }
```

Comment:
Since we're always doing the same `unset($_SESSION['messages'][$level][$key]);` for every conditional, it feels like this bit of code here could be improved consolidated.

Also, this feels like it should be a more generic/global duplicate message prevention functionality (in `backdrop_set_message()`, where we check whether the `$repeat` parameter is set to `TRUE`/`FALSE`), so that we can benefit in other cases/places from it.

Reply:
I've cleaned this up a little the latest PR.